### PR TITLE
Enable and disable controls

### PR DESCRIPTION
**Pull request by iangray001**
Allow controls to be enabled and disabled programmatically. The controls update visually to show their interactive status. 
Screenshot 2022-01-31 at 22 37 12

[README has been updated to document the API additions.](https://github.com/s00500/ESPUI/pull/154/commits/896dc97c751f9f143d828683a123cd422516d471#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5)

I did make a few implementation decisions here. The main one being that when disabled, a control will grey out, but the panel does not (it keeps its "Peterriver" colour etc.). This is mainly for flexibility and to support grouped controls. The README explains how to also grey the panel (as in the above screenshot) using Inline Styles.

Closes issue #56.

### CODE REVIEW

**Comment by nodoubtman** [comment]
@iangray001 thank you :)

File: data/js/controls.js [hunk 1]
```diff
@@ -639,27 +650,29 @@ function buttonclick(number, isdown) {
...
-      break;
-    case LEFT:
-      if (isdown) websock.send("pldown:" + number);
-      else websock.send("plup:" + number);
-      break;
-    case RIGHT:
-      if (isdown) websock.send("prdown:" + number);
-      else websock.send("prup:" + number);
-      break;
+  if(!$("#id" + number + " nav").hasClass("disabled")) {
```
Comment (s00500) [inline data/js/controls.js:653]: you could use an early return here :-)

**Review by s00500 — APPROVED** [review]
Cool stuff!

File: data/js/controls.js (hunk 1 above)
Comment (iangray001) [reply data/js/controls.js:653]: I'll fold this change into my next pr ;)

**Comment by thomastech** [comment]
@iangray001 
Hello Ian,

I installed the PR today and have been experiencing an odd problem. Randomly a control will become disabled. I didn't make any changes to my app, just updated the lib, recompiled my ESP32 project, and ran it.

I don't have a reliable way to invoke the bug, so background info is limited for now. At this point all I know is that it just happens randomly. So far I've had buttons and Text Inputs become disabled. Reboot and issue goes away, until the next time.

As I learn more I will let you know. If you have any things for me to try then ping me.

- Thomas

**Comment by iangray001** [comment]
So I've worked out the controls being disabled thing, I hadn't initialised the `enabled` value. That is in the latest commit on my repo.

The issue with clients getting out of sync actually isn't my fault ;) [This commit](https://github.com/iangray001/ESPUI/commit/f06781bc031856cb6dd8ba49c67f68406d2a985a) seems to be the problem. Specifically [this line](https://github.com/iangray001/ESPUI/commit/f06781bc031856cb6dd8ba49c67f68406d2a985a#r66823732). 

I'm not sure what is going on here, more investigation required, but I have reverted this line and it seems to be working again on my end.

**Comment by thomastech** [comment]
@iangray001 Many thanks! Feedback below:

1. PR #162 fixes the randomly disabled controls. 

2. Reverting the offending line #1006 in espui.cpp (from the Emulation PR #155) fixes the random client sync failure issue. I think this also resolved the random page timeouts I've been seeing (not previously reported, didn't want to load you up with more things to fix). But like my other recent ESPUI problems, this occurs randomly, so I can't confirm it is fully solved. Fingers crossed.

- Thomas

**Comment by iangray001** [comment]
I've thought more about that line and I now understand why it is a bug. The issue is that if clients come and go then IDs can be assigned with higher and higher IDs whilst the total count might remain low. For example you might end up with 2 clients, but with IDs 4 and 5. The committer has presumed that IDs will always be less than `count()`, and so clients fail to get updated.

I'm going to revert that, and a similar line elsewhere. I'll stick it on my existing PR.

**Comment by thomastech** [comment]
> The issue is that if clients come and go then IDs can be assigned with higher and higher IDs whilst the total count might remain low.

Hmm, that seems to explain one of my recent observations. Newly connected web clients would sync correctly while older web clients would often not remain in sync. But very random behavior since some controls would still sync. 

No two sessions were the same; I hate random bugs.

- Thomas
-

**Pull request merged**